src/plugins/fcrypt/fcrypt.c

```diff
@@ -0,0 +1,431 @@
+
+	if (!root) return 0;
+
+	// toplevel
+	if (strlen (keyString (root)) > 0)
+	{
+		recipientCount++;
+	}
+
+	ksRewind (config);
+	while ((k = ksNext (config)) != 0)
+	{
+		if (keyIsBelow (k, root))
+		{
+			recipientCount++;
+		}
+	}
+	return recipientCount;
+}
+
+/**
+ * @brief encrypt the file speicied at parentKey
+ * @param pluginConfig holds the plugin configuration
+ * @pararm parentKey holds the path to the file to be encrpted. Will hold an error description in case of failure.
+ * @retval 1 on success
+ * @retval -1 on error, errorKey holds an error description
+ */
+static int encrypt (KeySet * pluginConfig, Key * parentKey)
+{
+	const size_t recipientCount = getRecipientCount (pluginConfig);
+	if (recipientCount == 0)
+	{
+		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_CONFIG_FAULT, parentKey,
+				    "Missing GPG key (specified as %s) in plugin configuration.", ELEKTRA_CRYPTO_PARAM_GPG_KEY);
+		return -1;
+	}
+
+	const size_t testMode = inTestMode (pluginConfig);
+	int tmpFileFd = -1;
+
+	char * tmpFile = getTemporaryFileName (keyString (parentKey), &tmpFileFd);
+	if (!tmpFile)
+	{
+		ELEKTRA_SET_ERROR (87, parentKey, "Memory allocation failed");
+		return -1;
+	}
+
+	// prepare argument vector for gpg call
+	int argc = 8 + (2 * recipientCount) + (2 * testMode);
+	kdb_unsigned_short_t i = 0;
+	char * argv[argc];
+	argv[i++] = NULL;
+	argv[i++] = "--batch";
+	argv[i++] = "-o";
+	argv[i++] = tmpFile;
+	argv[i++] = "--yes"; // overwrite files if they exist
+
+	// add recipients
+	Key * k;
+	Key * root = ksLookupByName (pluginConfig, ELEKTRA_CRYPTO_PARAM_GPG_KEY, 0);
+
+	// append root (gpg/key) as gpg recipient
+	if (root && strlen (keyString (root)) > 0)
+	{
+		argv[i++] = "-r";
+		// NOTE argv[] values will not be modified, so const can be discarded safely
+		argv[i++] = (char *)keyString (root);
+	}
+
+	// append keys beneath root (crypto/key/#_) as gpg recipients
+	ksRewind (pluginConfig);
+	while ((k = ksNext (pluginConfig)) != 0)
+	{
+		if (keyIsBelow (k, root))
+		{
+			argv[i++] = "-r";
+			// NOTE argv[] values will not be modified, so const can be discarded safely
+			argv[i++] = (char *)keyString (k);
+		}
+	}
+
+	// if we are in test mode we add the trust model
+	if (testMode > 0)
+	{
+		argv[i++] = "--trust-model";
+		argv[i++] = "always";
+	}
+
+	// prepare rest of the argument vector
+	argv[i++] = "-e";
+	argv[i++] = (char *)keyString (parentKey);
+	argv[i++] = NULL;
+
+	// NOTE the encryption process works like this:
+	// gpg2 --batch --yes -o encryptedFile -r keyID -e configFile
+	// mv encryptedFile configFile
+
+	int result = CRYPTO_PLUGIN_FUNCTION (gpgCall) (pluginConfig, parentKey, NULL, argv, argc);
+
+	if (result == 1)
+	{
+		// encryption successful, overwrite the original file with the encrypted data
+		if (rename (tmpFile, keyString (parentKey)) != 0)
+		{
+			ELEKTRA_SET_ERRORF (31, parentKey, "Renaming file %s to %s failed.", tmpFile, keyString (parentKey));
+			result = -1;
+		}
```

Comment:
In case of success the tmp file is renamed to the "real" file in parentKey. How can I shred it?